{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "igDhE0gKOOiE"
      },
      "source": [
        "## RSALOR: Ready-to-Use Notebook\n",
        "[![PyPi Version](https://img.shields.io/pypi/v/rsalor.svg)](https://pypi.org/project/rsalor/) [![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT) [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](\n",
        "https://colab.research.google.com/github/3BioCompBio/RSALOR/blob/main/colab_notebook_RSALOR.ipynb)\n",
        "\n",
        "<img src=\"https://raw.githubusercontent.com/3BioCompBio/RSALOR/main/Logo.png\" height=\"250\" align=\"right\" style=\"height:200px;\">\n",
        "\n",
        "Ready-to-Use Notebook to run the RSALOR model:\n",
        " - Upload your MSA file\n",
        " - Uploae your 3D structure file\n",
        " - Run predictions on all single-site mutations\n",
        "\n",
        "The `rsalor` package combines structural data (Relative Solvent Accessibility, RSA) and evolutionary data (Log Odd Ratio, LOR from MSA) to evaluate effects of missense mutations in proteins.\n",
        "It computes the `RSA*LOR` score for each single-site missense mutation in a target protein by combining multiple computational steps into a fast and user-friendly tool.\n",
        "Source code in the [RSALOR GitHub](https://github.com/3BioCompBio/RSALOR).\n",
        "\n",
        "**Please cite**:\n",
        "- [Matsvei Tsishyn, Pauline Hermans, Fabrizio Pucci, Marianne Rooman (2025). Residue conservation and solvent accessibility are (almost) all you need for predicting mutational effects in proteins. Bioinformatics, btaf322](https://doi.org/10.1093/bioinformatics/btaf322).\n",
        "\n",
        "- [Pauline Hermans, Matsvei Tsishyn, Martin Schwersensky, Marianne Rooman, Fabrizio Pucci (2024). Exploring evolution to uncover insights into protein mutational stability. Molecular Biology and Evolution, 42(1), msae267](https://doi.org/10.1093/molbev/msae267).\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "yzWtI1teMTTA"
      },
      "outputs": [],
      "source": [
        "#@title Install and import dependencies\n",
        "\n",
        "# Install packages -------------------------------------------------------------\n",
        "!pip install rsalor\n",
        "\n",
        "\n",
        "# Import packages --------------------------------------------------------------\n",
        "import os\n",
        "import requests\n",
        "from google.colab import files\n",
        "from Bio.PDB import PDBParser, PPBuilder, PDBList\n",
        "from rsalor import MSA\n",
        "from rsalor.sequence import Sequence, FastaReader, PairwiseAlignment\n",
        "from rsalor.structure import Structure\n",
        "\n",
        "\n",
        "# Dependencies (small helper functions) ----------------------------------------\n",
        "\n",
        "# Log functions\n",
        "def clip_string(input_str: str, max_len: int=100) -> str:\n",
        "  \"\"\"Truncate a string and append '...' if it exceeds max_len.\"\"\"\n",
        "  return input_str if len(input_str) <= max_len else input_str[:max_len] + \"...\"\n",
        "\n",
        "# Fetch functions\n",
        "def is_valid_pdb_id(pdb_id :str) -> bool:\n",
        "  \"\"\"Returns if 'input' can be a PDB-ID.\"\"\"\n",
        "  POSSIBLE_FIRST_CHARACTERS = \"123456789\"\n",
        "  POSSIBLE_FOLLOWING_CHARACTERS = \"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ\"\n",
        "  input_upper = pdb_id.upper()\n",
        "  return len(input_upper) == 4 \\\n",
        "    and input_upper[0] in POSSIBLE_FIRST_CHARACTERS \\\n",
        "    and all([char in POSSIBLE_FOLLOWING_CHARACTERS for char in input_upper[1:]])\n",
        "\n",
        "def fetch_pdb(pdb_id: str) -> str:\n",
        "  \"\"\"Fetch a '.pdb' file from the PDB and return its path.\"\"\"\n",
        "  pdb_id = pdb_id.lower().strip()\n",
        "  if not is_valid_pdb_id(pdb_id):\n",
        "    raise ValueError(f\"❌ pdb_id='{pdb_id}' is not a valid PDB Id.\")\n",
        "  pdb_fetcher = PDBList()\n",
        "  file_path = pdb_fetcher.retrieve_pdb_file(pdb_id, file_format=\"pdb\", pdir=\"./\")\n",
        "  if file_path is None:\n",
        "    raise ValueError(f\"❌ Fetch pdb_id='{pdb_id}' has failed.\")\n",
        "  if file_path.endswith(\".ent\"):\n",
        "    file_path_old = file_path\n",
        "    file_path = file_path.removesuffix(\".ent\") + \".pdb\"\n",
        "    os.rename(file_path_old, file_path)\n",
        "  return file_path\n",
        "\n",
        "def fetch_uniprot(uniprot_id: str) -> str:\n",
        "  \"\"\"Fetch a '.pdb' file from the AlphaFoldDB by its UniProt ID and return its path.\"\"\"\n",
        "  uniprot_id = uniprot_id.upper().strip()\n",
        "  url = f\"https://alphafold.ebi.ac.uk/files/AF-{uniprot_id}-F1-model_v6.pdb\"\n",
        "  file_path = f\"./{uniprot_id}.pdb\"\n",
        "  r = requests.get(url)\n",
        "  if r.status_code != 200:\n",
        "    raise ValueError(f\"❌ AlphaFoldDB fetch failed for UniProt ID '{uniprot_id}' from '{url}'\")\n",
        "  with open(file_path, \"wb\") as fs:\n",
        "    fs.write(r.content)\n",
        "  return file_path"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "iKcd15dHQySw"
      },
      "outputs": [],
      "source": [
        "#@title Upload MSA file (in '.fasta', '.a2m' or '.a3m' format)\n",
        "\n",
        "# Drag-and-drop file picker\n",
        "uploaded = files.upload()\n",
        "msa_path: str = list(uploaded.keys())[0]\n",
        "\n",
        "# Guardian for correct extension\n",
        "if not any(msa_path.endswith(ext) for ext in MSA.ACCEPTED_EXTENTIONS):\n",
        "    raise ValueError(f\"❌ Uploaded MSA file '{msa_path}' should have extention among {MSA.ACCEPTED_EXTENTIONS}.\")\n",
        "\n",
        "# Log uploaded MSA file\n",
        "print(f\" * ✅ MSA file uploaded: '{msa_path}'\")\n",
        "taget_sequence = FastaReader.read_first_sequence(msa_path)\n",
        "taget_sequence_str = taget_sequence.sequence if len(taget_sequence) < 100 else taget_sequence.sequence[:100] + \"...\"\n",
        "print(f\" * ✅ taget sequence (L={len(taget_sequence)}): '{clip_string(taget_sequence_str)}'\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "TjEwK9t9t6E0"
      },
      "outputs": [],
      "source": [
        "#@title Upload 3D structure file (in '.pdb' format)\n",
        "\n",
        "# Define upload method ---------------------------------------------------------\n",
        "#@markdown ### Choose your 3D structure upload method (upload file or fetch from the PDB or from the AlphaFold-DB)\n",
        "upload_method = \"upload_local_file\" # @param [\"upload_local_file\",\"fetch_from_pdb\",\"fetch_from_alphafold_db\"]\n",
        "retrieval_id = \"\" # @param {\"type\":\"string\",\"placeholder\":\"PDB ID like '6m0j' or UniProt ID like 'Q9LW00'\"}\n",
        "#@markdown - if using `upload_local_file`, leave empty\n",
        "#@markdown - if using `fetch_from_pdb`, specify the **PDB ID** (like `6m0j`)\n",
        "#@markdown - if using `fetch_from_alphafold_db`, specify the **UniProt ID** (like `Q9LW00`)\n",
        "\n",
        "# Case: Drag-and-drop local file -----------------------------------------------\n",
        "if upload_method == \"upload_local_file\":\n",
        "\n",
        "  # Drag-and-drop file picker\n",
        "  uploaded = files.upload()\n",
        "  pdb_path: str = list(uploaded.keys())[0]\n",
        "\n",
        "  # Guardian for correct extension\n",
        "  if not pdb_path.endswith(\".pdb\"):\n",
        "      raise ValueError(f\"❌ Uploaded PDB file '{pdb_path}' should have extention '.pdb'.\")\n",
        "  pdb_name = pdb_path.removesuffix(\".pdb\")\n",
        "  print(f\" * ✅ PDB file uploaded: '{pdb_path}'\")\n",
        "\n",
        "# Case: fetch file from the PDB ------------------------------------------------\n",
        "elif upload_method == \"fetch_from_pdb\":\n",
        "\n",
        "  # retrieval_id not fill error\n",
        "  if retrieval_id == \"\" or retrieval_id is None:\n",
        "    raise ValueError(f\"❌ If upload_method='{upload_method}', please specify a retrieval_id.\")\n",
        "\n",
        "  # Fetch PDB file\n",
        "  pdb_path: str = fetch_pdb(retrieval_id)\n",
        "\n",
        "  # Log\n",
        "  pdb_name = pdb_path.removesuffix(\".pdb\")\n",
        "  print(f\" * ✅ PDB file fetched: '{pdb_path}'\")\n",
        "\n",
        "# Case: fetch file from the AlphaFold-DB ---------------------------------------\n",
        "elif upload_method == \"fetch_from_alphafold_db\":\n",
        "\n",
        "  # retrieval_id not fill error\n",
        "  if retrieval_id == \"\" or retrieval_id is None:\n",
        "    raise ValueError(f\"❌ If upload_method='{upload_method}', please specify a retrieval_id.\")\n",
        "\n",
        "  # Fetch PDB file\n",
        "  pdb_path: str = fetch_uniprot(retrieval_id)\n",
        "\n",
        "  # Log\n",
        "  pdb_name = pdb_path.removesuffix(\".pdb\")\n",
        "  print(f\" * ✅ AF-DB file fetched: '{pdb_path}'\")\n",
        "\n",
        "# Case: error ------------------------------------------------------------------\n",
        "else:\n",
        "  raise ValueError(f\"❌ Unknown upload_method='{upload_method}'.\")\n",
        "\n",
        "# Validate and Log -------------------------------------------------------------\n",
        "# Parse PDB file\n",
        "pp_builer = PPBuilder()\n",
        "structure = PDBParser(QUIET=True).get_structure(\"protein\", pdb_path)\n",
        "\n",
        "# Log uploaded PDB file\n",
        "sequences_by_chain: dict[str, Sequence] = {}\n",
        "for chain in structure[0]: # loop only on model 1\n",
        "  aa_seq_segments: list[str] = pp_builer.build_peptides(chain)\n",
        "  aa_seq: str = \"\".join([str(pp.get_sequence()) for pp in aa_seq_segments]) # concatenate fragments if multiple\n",
        "  print(f\"    - chain {chain.id} (L={len(aa_seq)}): '{clip_string(aa_seq)}'\")\n",
        "  sequences_by_chain[chain.id] = Sequence(f\"{pdb_name}_{chain.id}\", aa_seq)\n",
        "print(f\" * ✅ Choose target chain among {len(sequences_by_chain)} detected chain(s): '{''.join(sequences_by_chain.keys())}'\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "yTKvX4WCz9W9"
      },
      "outputs": [],
      "source": [
        "#@title Select PDB chain\n",
        "chain = \"A\" # @param {\"type\":\"string\",\"placeholder\":\"A\"}\n",
        "\n",
        "# Guardians\n",
        "if len(chain) != 1:\n",
        "  raise ValueError(f\"❌ chain='{chain}' should be a string of length 1.\")\n",
        "if chain not in sequences_by_chain:\n",
        "  raise ValueError(f\"❌ chain='{chain}' not in PDB '{pdb_path}' (among '{''.join(sequences_by_chain.keys())}')\")\n",
        "\n",
        "# Show alignments\n",
        "print(f\" * ✅ MSA target sequence aligned to chain '{chain}' in PDB structure.\")\n",
        "align = PairwiseAlignment(taget_sequence, sequences_by_chain[chain])\n",
        "align.show();\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "Ebp05L4F7p3g"
      },
      "outputs": [],
      "source": [
        "#@title Run RSALOR\n",
        "\n",
        "# Output settings\n",
        "#@markdown ### Output settings\n",
        "output_name = \"\" # @param {\"type\":\"string\",\"placeholder\":\"<msa_name>_rsalor\"}\n",
        "#@markdown - leave empty for auto\n",
        "sep = \",\" # @param {\"type\":\"string\",\"placeholder\":\",\"}\n",
        "#@markdown - separator in output CSV file\n",
        "\n",
        "# RSALOR run settings\n",
        "#@markdown ### RSALOR settings\n",
        "theta_regularization = 0.01 # @param {\"type\":\"number\",\"placeholder\":\"0.01\"}\n",
        "#@markdown - regularization term for LOR/LR at amino acid frequencies level\n",
        "seqid_weights = 0.80 # @param {\"type\":\"number\",\"placeholder\":\"0.80\"}\n",
        "#@markdown - seqid threshold to consider two sequences in the same cluster for weighting (set None to ignore)\n",
        "min_seqid = 0.35 # @param {\"type\":\"number\",\"placeholder\":\"0.35\"}\n",
        "#@markdown - discard sequences which seqid with target sequence is below (set None to ignore)\n",
        "num_threads = 2 # @param {\"type\":\"integer\",\"placeholder\":\"2\"}\n",
        "#@markdown - number of threads (CPUs) for weights evaluation (in the C++ backend)\n",
        "\n",
        "# Run RSALOR\n",
        "msa = MSA(\n",
        "    msa_path, pdb_path, chain,\n",
        "    theta_regularization=theta_regularization,\n",
        "    seqid_weights=seqid_weights,\n",
        "    min_seqid=min_seqid,\n",
        "    num_threads=num_threads,\n",
        "    verbose=True,\n",
        ")\n",
        "\n",
        "# Set output path\n",
        "if output_name == \"\" or output_name is None:\n",
        "  output_name = f\"{msa.name}_rsalor\"\n",
        "output_path = f\"{output_name}.csv\"\n",
        "\n",
        "# Compute and save scores\n",
        "rsalor_scores = msa.save_scores(\n",
        "    output_path,\n",
        "    sep=sep,\n",
        "    round_digit=6,\n",
        "    log_results=True,\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KmW_SkOqEGuG"
      },
      "outputs": [],
      "source": [
        "#@title Download RSALOR output\n",
        "files.download(output_path)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
